File: frontend/web/src/api/subscriptions.ts
/**
 * 策略订阅 API
 */
import client from './client';

// 类型定义
export interface Subscription {
  id: number;
  userId: number;
  strategyId: number;
  strategyName: string;
  stockPoolType: 'all' | 'group' | 'custom';
  stockGroupId?: number;
  customSymbols?: string[];
  notifyEnabled: boolean;
  notifyChannels: string[];
  stockPoolCount: number;
  createdAt: string;
  updatedAt?: string;
}

export interface StockGroup {
  id: number;
  userId: number;
  name: string;
  description?: string;
  isDefault: boolean;
  stockCount: number;
  createdAt: string;
  updatedAt?: string;
}

// 获取订阅列表
export function getSubscriptions(strategyId?: number) {
  const params = strategyId ? { strategy_id: strategyId } : {};
  return client.get<{ status: string; data: Subscription[] }>('/subscriptions', { params });
}

// 创建订阅
export function createSubscription(data: {
  strategyId: number;
  stockPoolType?: 'all' | 'group' | 'custom';
  stockGroupId?: number;
  customSymbols?: string[];
  notifyEnabled?: boolean;
  notifyChannels?: string[];
}) {
  return client.post<{ status: string; data: Subscription }>('/subscriptions', data);
}

// 更新订阅
export function updateSubscription(
  subscriptionId: number,
  data: {
    stockPoolType?: 'all' | 'group' | 'custom';
    stockGroupId?: number;
    customSymbols?: string[];
    notifyEnabled?: boolean;
    notifyChannels?: string[];
  }
) {
  return client.put<{ status: string; data: Subscription }>(`/subscriptions/${subscriptionId}`, data);
}

// 删除订阅
export function deleteSubscription(subscriptionId: number) {
  return client.delete<{ status: string }>(`/subscriptions/${subscriptionId}`);
}

// 获取订阅详情
export function getSubscription(subscriptionId: number) {
  return client.get<{ status: string; data: Subscription }>(`/subscriptions/${subscriptionId}`);
}

// 获取自选分组列表（订阅页面用）
export function getGroups() {
  return client.get<{ status: string; data: StockGroup[] }>('/favorites/groups');
}

File: backend/app/api/endpoints/subscriptions.py
"""
策略订阅 API 端点 - Subscriptions API
功能：策略订阅的 CRUD 操作
"""
from fastapi import APIRouter, HTTPException, Depends, Query
from typing import Optional, List
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.services.subscription_service import SubscriptionService

router = APIRouter()


# ==================== Pydantic 模型 ====================

class SubscriptionCreate(BaseModel):
    """创建订阅请求模型"""
    strategyId: int = Field(..., description="策略ID")
    stockPoolType: str = Field("all", description="股票池类型: all/group/custom")
    stockGroupId: Optional[int] = Field(None, description="分组ID（当stockPoolType为group时使用）")
    customSymbols: Optional[List[str]] = Field(None, description="自定义股票列表（当stockPoolType为custom时使用）")
    notifyEnabled: bool = Field(True, description="是否启用通知")
    notifyChannels: Optional[List[str]] = Field(["websocket"], description="通知渠道")


class SubscriptionUpdate(BaseModel):
    """更新订阅请求模型"""
    stockPoolType: Optional[str] = Field(None, description="股票池类型: all/group/custom")
    stockGroupId: Optional[int] = Field(None, description="分组ID")
    customSymbols: Optional[List[str]] = Field(None, description="自定义股票列表")
    notifyEnabled: Optional[bool] = Field(None, description="是否启用通知")
    notifyChannels: Optional[List[str]] = Field(None, description="通知渠道")


# ==================== 订阅管理 ====================

@router.get("", summary="获取订阅列表", tags=["策略订阅"])
async def get_subscriptions(
    strategy_id: Optional[int] = Query(None, description="策略ID筛选"),
    user_id: int = 1,
    db: Session = Depends(get_db)
):
    """获取用户订阅列表"""
    try:
        service = SubscriptionService(db)
        subscriptions = service.get_subscriptions(user_id, strategy_id)

        return {
            "status": "success",
            "data": subscriptions
        }

    except Exception as e:
        raise HTTPException(status_code=500, detail=f"获取订阅列表失败: {str(e)}")


@router.post("", summary="创建订阅", tags=["策略订阅"])
async def create_subscription(
    request: SubscriptionCreate,
    user_id: int = 1,
    db: Session = Depends(get_db)
):
    """订阅策略"""
    try:
        service = SubscriptionService(db)
        subscription = service.create_subscription(
            user_id=user_id,
            strategy_id=request.strategyId,
            stock_pool_type=request.stockPoolType,
            stock_group_id=request.stockGroupId,
            custom_symbols=request.customSymbols,
            notify_enabled=request.notifyEnabled,
            notify_channels=request.notifyChannels
        )

        return {
            "status": "success",
            "message": "订阅成功",
            "data": subscription
        }

    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/{subscription_id}", summary="获取订阅详情", tags=["策略订阅"])
async def get_subscription(
    subscription_id: int,
    user_id: int = 1,
    db: Session = Depends(get_db)
):
    """获取订阅详情"""
    try:
        service = SubscriptionService(db)
        subscription = service.get_subscription(subscription_id, user_id)

        if not subscription:
            raise HTTPException(status_code=404, detail=f"订阅不存在: {subscription_id}")

        return {
            "status": "success",
            "data": subscription
        }

    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"获取订阅详情失败: {str(e)}")


@router.put("/{subscription_id}", summary="更新订阅", tags=["策略订阅"])
async def update_subscription(
    subscription_id: int,
    request: SubscriptionUpdate,
    user_id: int = 1,
    db: Session = Depends(get_db)
):
    """更新订阅设置"""
    try:
        service = SubscriptionService(db)
        subscription = service.update_subscription(
            subscription_id=subscription_id,
            user_id=user_id,
            stock_pool_type=request.stockPoolType,
            stock_group_id=request.stockGroupId,
            custom_symbols=request.customSymbols,
            notify_enabled=request.notifyEnabled,
            notify_channels=request.notifyChannels
        )

        return {
            "status": "success",
            "message": "订阅更新成功",
            "data": subscription
        }

    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.delete("/{subscription_id}", summary="取消订阅", tags=["策略订阅"])
async def delete_subscription(
    subscription_id: int,
    user_id: int = 1,
    db: Session = Depends(get_db)
):
    """取消订阅"""
    try:
        service = SubscriptionService(db)
        service.delete_subscription(subscription_id, user_id)

        return {
            "status": "success",
            "message": "取消订阅成功"
        }

    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
